### PR TITLE
Working towards v1.0 Release

### PR DESCRIPTION
**Pull request by JosXa**

### CODE REVIEW

File: tgintegration/collector.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    """
+
+
+class NotSet(object):
+    pass  # TODO: Use a better object for this
+
+
+@dataclass
+class Expectation:
+    min_messages: Union[int, NotSet] = NotSet
```
Comment (joshcoales) [inline tgintegration/collector.py:47]: `NotSet` is not an instance of `NotSet`, it's an instance of `Type[NotSet]`

File: tgintegration/collector.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    This can cause the total duration to exceed the `max_wait` time.
+    """
+
+    raise_on_timeout: bool = False
+    """
+    Whether to raise an exception when a timeout occurs or to fail with a log message.
+    """
+
+
+class NotSet(object):
```
Comment (joshcoales) [inline tgintegration/collector.py:41]: Might be best to do a sentinel value in the way dataclasses does:
https://github.com/python/cpython/blob/master/Lib/dataclasses.py#L169

I.e.
```
class _NotSetType:
    pass
NOT_SET = _NotSetType()

...

class Expectation:
    min_messages: Union[int, _NotSetType] = NOT_SET
```

File: tgintegration/collector.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    async with add_handler_transient(controller.client, handler):
+        response = Response(controller, recorder)
+
+        logger.debug("Collector set up. Executing user-defined interaction...")
+        yield response  # Start user-defined interaction
+        logger.debug("interaction complete.")
+
+        num_received = 0
+        last_received_timestamp = (
+            None  # TODO: work with the message's timestamp instead of utcnow()
```
Comment (joshcoales) [inline tgintegration/collector.py:110]: Another TODO to implement

File: tgintegration/collector.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+                    logger.debug(f"Waiting for message #{num_received + 1}")
+                    await asyncio.wait_for(
+                        recorder.wait_until(
+                            lambda msgs: expectation.is_sufficient(msgs)
+                            or len(msgs) > num_received
+                        ),
+                        timeout=seconds_remaining,
+                    )
+
+                num_received = len(recorder.messages)  # TODO: this is ugly
```
Comment (joshcoales) [inline tgintegration/collector.py:129]: Why is it ugly?

File: tgintegration/containers/inlineresults.py [hunk 5]
```diff
@@ -131,7 +133,11 @@ def _match(self, pattern, getter) -> List:
...
     def find_results(
-        self, title_pattern=None, description_pattern=None, message_pattern=None, url_pattern=None,
+        self,
+        title_pattern=None,
+        description_pattern=None,
+        message_pattern=None,
+        url_pattern=None,
     ) -> Set[InlineResult]:
 
         # TODO:
```
Comment (joshcoales) [inline tgintegration/containers/inlineresults.py:143]: Still to do?

File: tgintegration/containers/keyboard.py [hunk 6]
```diff
@@ -1,91 +1,88 @@
...
     def __init__(
         self,
-        client: Client,
+        controller: "BotController",
         chat_id: Union[int, str],
         message_id: int,
         button_rows: List[List[KeyboardButton]],
     ):
-        self._client = weakref.proxy(client)
+        self._controller: BotController = weakref.proxy(controller)
```
Comment (joshcoales) [inline tgintegration/containers/keyboard.py:22]: Why a weakref?

File: tgintegration/containers/response.py [hunk 7]
```diff
@@ -108,12 +101,12 @@ def last_message_timestamp(self) -> Optional[datetime]:
         if self.empty:
             return None
         # TODO: Dan should fix this
```
Comment (joshcoales) [inline tgintegration/containers/response.py:103]: Who is Dan?

**Review by joshcoales — COMMENTED** [review]
This looks great! Got a few comments here and there, a lot of them are about lingering #TODO comments, and requests for some pydoc to explain some parameters. But looks like a great architecture.

I would love an example of testing a bot in a group? Hopefully that's possible after the rework?

File: .idea/BlackPycharmConfig.xml [hunk 8]
```diff
@@ -3,4 +3,4 @@
   <component name="BlackPycharmConfig">
     <option name="executableName" value="C:\Program Files\Python36\Scripts\black.exe" />
   </component>
-</project>
\ No newline at end of file
+</project>
```
[the remark this answers is not in the document]
Comment (JosXa) [reply .idea/BlackPycharmConfig.xml:6]: Nnnnope :D good catch

File: .travis.yml [hunk 9]
```diff
@@ -1,27 +0,0 @@
-# Config file for automatic testing at travis-ci.org
```
[the remark this answers is not in the document]
Comment (JosXa) [reply .travis.yml:1]: File is removed 👍

File: docs/conf.py [hunk 10]
```diff
@@ -210,9 +207,13 @@
...
 # [howto/manual]).
 latex_documents = [
-    ('index', 'tgintegration.tex',
-     u'tgintegration Documentation',
-     u'Joscha Götzer', 'manual'),
+    (
+        "index",
+        "tgintegration.tex",
+        u"tgintegration Documentation",
+        u"Joscha Götzer",
```
[the remark this answers is not in the document]
Comment (JosXa) [reply docs/conf.py:214]: File is removed 👍

File: examples/pytest/conftest.py [hunk 11]
```diff
@@ -1,51 +1,88 @@
...
-    # event_loop.run_until_complete(controller.start())
-    print("Client ready.")
-
-    yield controller  # py.test sugar to separate setup from teardown
 
-    # Teardown
-    await controller.stop()
-    # event_loop.run_until_complete(controller.stop())
+# @pytest.yield_fixture(scope="session")
+# async def controller(event_loop):
```
[the remark this answers is not in the document]
Comment (JosXa) [reply examples/pytest/conftest.py:57]: I suppose, yes. Git doesn't forget ;)

File: tgintegration/containers/inlineresults.py (hunk 5 above)
Comment (JosXa) [reply tgintegration/containers/inlineresults.py:143]: Yep, I haven't touched the inline stuff yet.

- [ ] touch the inline stuff 😄

File: tgintegration/containers/keyboard.py (hunk 6 above)
Comment (JosXa) [reply tgintegration/containers/keyboard.py:22]: My old me from two years ago didn't know what the heck he was doing. Good catch, removing!

File: tgintegration/containers/response.py (hunk 7 above)
Comment (JosXa) [reply tgintegration/containers/response.py:103]: https://t.me/haskell <-- Pyrogram author

File: examples/pytest/conftest.py [hunk 12]
```diff
@@ -1,51 +1,88 @@
...
+
+
+logging.getLogger("pyrogram").setLevel(logging.WARNING)
 
 
 @pytest.yield_fixture(scope="session")
-def event_loop():
-    loop = asyncio.get_event_loop()
+def event_loop(request):
+    """Create an instance of the default event loop for each test case."""
```
[the remark this answers is not in the document]
Comment (JosXa) [reply examples/pytest/conftest.py:22]: Good catch, someone's reading comments! :D

File: tgintegration/collector.py (hunk 4 above)
Comment (JosXa) [reply tgintegration/collector.py:129]: - [ ] Add more info about a better architecture

File: tgintegration/containers/response.py (hunk 7 above)
Comment (joshcoales) [reply tgintegration/containers/response.py:103]: Ah, okay

**Comment by JosXa** [comment]
self-high-five ✋  #hacktoberfest-accepted

**Pull request merged**